File: app/frontend/src/types/product.ts
/**
 * Domain types for the catalogue. These mirror the contract the FastAPI backend
 * will expose, so the frontend can switch from mock data to the real API with no
 * component changes. Keep these in sync with the backend Pydantic schemas in v2.
 */

/** Product categories. Open-ended by design — new categories ship in future releases. */
export type ProductCategory = 'tshirts' | 'shirts' | 'hoodies' | 'jackets';

export interface ProductSize {
  /** e.g. "S", "M", "L", "XL" */
  label: string;
  /** Whether this size is currently purchasable. */
  inStock: boolean;
}

export interface ProductImage {
  url: string;
  alt: string;
}

export interface Product {
  /** Stable identifier (slug-friendly). */
  id: string;
  slug: string;
  name: string;
  category: ProductCategory;
  /** Short marketing line shown on cards. */
  tagline: string;
  /** Long description shown on the product page. */
  description: string;
  /** Price in integer minor units (cents) to avoid float errors. */
  priceCents: number;
  currency: string;
  images: ProductImage[];
  sizes: ProductSize[];
  /** Surface badges like "New" or "Last few". */
  badge?: string;
  /** Featured products appear on the home page. */
  featured?: boolean;
}

export const CATEGORY_LABELS: Record<ProductCategory, string> = {
  tshirts: 'T-Shirts',
  shirts: 'Shirts',
  hoodies: 'Hoodies',
  jackets: 'Jackets',
};

/** Ordered list used to render filters/nav consistently. */
export const CATEGORY_ORDER: ProductCategory[] = ['tshirts', 'shirts', 'hoodies', 'jackets'];

File: app/backend/catalog-service/app/schemas/product.py
"""API (Pydantic) schemas for the catalogue.

These define the wire contract consumed by the frontend (`src/types/product.ts`)
and the order service. Field names are camelCase on the wire to match the
TypeScript client, while the ORM uses snake_case — bridged via aliases.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class CamelModel(BaseModel):
    """Base that serialises to camelCase and accepts either case on input."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        from_attributes=True,
    )


class ProductImage(CamelModel):
    url: str
    alt: str


class ProductSize(CamelModel):
    label: str
    in_stock: bool = True


class ProductRead(CamelModel):
    id: str
    slug: str
    name: str
    category: str
    tagline: str
    description: str
    price_cents: int = Field(..., ge=0)
    currency: str
    images: list[ProductImage]
    sizes: list[ProductSize]
    badge: str | None = None
    featured: bool = False
